File: leetcode/medium/208_trie/trie.py
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.children = {}
        self.end_of_word = False

    def addChild(self, child):
        self.children[child.val] = child

    def getChild(self, val):
        if val not in self.children:
            return None
        return self.children[val]


class Trie:
    def __init__(self):
        self.root = Node("")

    def insert(self, word: str) -> None:
        self.insertRec(self.root, word)

    def insertRec(self, node, word):
        if len(word) == 0:
            node.end_of_word = True
            return
        child = node.getChild(word[0])
        if not child:
            child = Node(word[0])
            node.addChild(child)
        self.insertRec(child, word[1:])

    def search(self, word: str) -> bool:
        return self.searchRec(self.root, word)

    def searchRec(self, node, word):
        if len(word) == 0:
            return node.end_of_word
        child = node.getChild(word[0])
        if not child:
            return False
        return self.searchRec(child, word[1:])

    def startsWith(self, prefix: str) -> bool:
        return self.startsWithRec(self.root, prefix)

    def startsWithRec(self, node, prefix):
        if len(prefix) == 0:
            return True
        child = node.getChild(prefix[0])
        print(child)
        if not child:
            print(child)
            return False
        return self.startsWithRec(child, prefix[1:])


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: leetcode/medium/208_trie/trie.cpp
#include <string>
#include <vector>

class TrieNode {
public:
  TrieNode(char key) : key(key), children(), end_of_word(false) {};
  char key;
  bool end_of_word;
  std::vector<TrieNode *> children;
};

class Trie {
public:
  Trie() { root = new TrieNode(0); }
  ~Trie() { destroy(root); }

  void insert(std::string word) {
    auto node = root;
    for (const auto c : word) {
      printf("%c\n", c);
      TrieNode *loc = nullptr;
      for (auto child : node->children) {
        if (child->key == c) {
          loc = child;
          break;
        }
      }
      if (loc == nullptr) {
        TrieNode *new_node = new TrieNode(c);
        node->children.push_back(new_node);
        node = new_node;
      } else {
        node = loc;
      }
    }
    node->end_of_word = true;
  }

  bool search(std::string word) {
    auto node = root;
    for (const auto c : word) {
      TrieNode *loc = nullptr;
      for (auto child : node->children) {
        if (child->key == c) {
          loc = child;
          break;
        }
      }
      if (loc == nullptr) {
        return false;
      }
      node = loc;
    }
    return node->end_of_word;
  }

  bool startsWith(std::string prefix) {
    auto node = root;
    for (const auto c : prefix) {
      TrieNode *loc = nullptr;
      for (auto child : node->children) {
        if (child->key == c) {
          loc = child;
          break;
        }
      }
      if (loc == nullptr) {
        return false;
      }
      node = loc;
    }
    return true;
  }

private:
  void destroy(TrieNode *root) {
    for (auto child : root->children) {
      if (child) {
        destroy(child);
      }
    }
    delete root;
  }

  TrieNode *root;
};
